File: LeetCode_Python/Test112.py
__author__ = 'zyt'

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @param {integer} sum
    # @return {boolean}
    def hasPathSum(self, root, sum):
        if not root:
            return False
        if not root.left and not root.right:
            return root.val == sum
        return self.hasPathSum(root.left, sum - root.val) or self.hasPathSum(root.right, sum - root.val)
        # left = right = False
        # if root.left:
        #     left = self.hasPathSum(root.left, sum - root.val)
        # if root.right:
        #     right = self.hasPathSum(root.right, sum - root.val)
        # return left or right

File: LeetCode_Java/src/Test112.java
/**
 * Created by zyt on 15/8/19 20:30.
 */
public class Test112 {
    /**
     * Definition for a binary tree node.
     */
    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    private class Solution {
        public boolean hasPathSum(TreeNode root, int sum) {
            if (root == null) {
                return false;
            }
            if (root.left == null && root.right == null && root.val == sum) {
                return true;
            }
            boolean left = this.hasPathSum(root.left,sum-root.val);
            boolean right = this.hasPathSum(root.right, sum - root.val);
            return left || right;
        }
    }
}
